Refactor Config Package

eliasjf opened the pull request:
This PR refactors the config package to remove the `idp.LdapIDP` field. Part of those changes include reworking how a `tls.Config` is generated. This refactor is intended to setup the upcoming CONNECTOR work.

eliasjf commented on pkg/config/config.go:
```diff
@@ -96,28 +103,85 @@ func (cm *ConfigManager) ReadConfig(filename string) (*Config, error) {
+
+// processOutboundTLS builds a *tls.Config for outbound connections to remote services. This *tls.Config can be used by
+// connectors to authenticate against a target IDP.
+func processOutboundTLS(logger log.Logger, fs afero.Fs, conf TLS) (*tls.Config, error) {
```
This method was ported over from `ldap_config.go`. I changed the name and slightly changed the signature.

eliasjf commented on pkg/config/config.go:
```diff
@@ -27,14 +25,23 @@ var loadX509KeyPair = func(certFile, keyFile string) (tls.Certificate, error) {
 type TLS struct {
+	// InboundTLS is a flag to specify whether the TLS config is intended for inbound connections. Inbound TLS is used
+	// for client authentication while outbound TLS is used to authenticate against a remote service.
+	InboundTLS bool `yaml:"inboundTLS"`
```
I expect we can ultimately remove this flag. I didn't want to get bogged down with TLS specific work and saw this as the cleanest way to minimize changes.

rmulley replied:
Agree, I wonder if the structure even needs this field and can just populate what is provided. The configuration could dictate whether it's inbound or outbound.

topherMarie commented on pkg/config/config.go:
```diff
@@ -27,14 +25,23 @@ var loadX509KeyPair = func(certFile, keyFile string) (tls.Certificate, error) {
-	LDAP          idp.LdapIDP `yaml:"ldap"`
-	TLS           TLS         `yaml:"tls"`
+	ListenAddress string `yaml:"listenAddress"`
+	Name          string
```
What is the Name for in here?

rmulley commented on pkg/config/config.go:
```diff
@@ -96,28 +103,85 @@ func (cm *ConfigManager) ReadConfig(filename string) (*Config, error) {
+
+	// Only attempt to load root CAs if configuration data was provided.
+	var rootCAs *x509.CertPool
+	if conf.CAFile != "" {
```
I wonder if this block could be combined with `buildClientCACertPool`, provided the method was made slightly more generic.

rmulley replied:
We don't have to do this now if we don't want to.

eliasjf commented:
@topherMarie, @rmulley, I've updated this PR to:
1. Remove the changes I had previously made to the LDAP connector and corresponding example. This will help avoid a nasty merge conflict since Robert is working against the LDAP connector.
2. Reworked the config package to add the concept of `Connectors` and added a "dummy" connector and corresponding example to show this in action.

You'll notice that the tests are currently failing. This is because the config changes break both `examples/ldap` and `cmd/maverics`. I of course plan to fix these programs but I think it will be useful to get some interim feedback and to possibly make the fixes in another PR.
